presto-main/src/test/java/com/facebook/presto/execution/scheduler/group/TestDynamicLifespanScheduler.java

```diff
@@ -0,0 +1,231 @@
+            lifespanScheduler.onLifespanExecutionFinished(sourceScheduler.getLastStartedLifespans());
+            assertEquals(sourceScheduler.getLastStartedLifespans().size(), 1);
+            sourceScheduler.getLastStartedLifespans().clear();
+        }
+    }
+
+    @Test
+    public void testAllTaskFailed()
+    {
+        LifespanScheduler lifespanScheduler = getLifespanScheduler();
+        TestSourceScheduler sourceScheduler = new TestSourceScheduler();
+        lifespanScheduler.scheduleInitial(sourceScheduler);
+        assertEquals(sourceScheduler.getLastStartedLifespans().size(), 2);
+
+        assertTrue(lifespanScheduler.retryFailedTask(0, ImmutableList.of(sourceScheduler)));
+        assertFalse(lifespanScheduler.retryFailedTask(1, ImmutableList.of(sourceScheduler)));
+        assertThrows(IllegalStateException.class, () -> lifespanScheduler.schedule(sourceScheduler));
+    }
+
+    private static LifespanScheduler getLifespanScheduler()
+    {
+        return new DynamicLifespanScheduler(
+                new DynamicBucketNodeMap(split -> ((TestSplit) split.getConnectorSplit()).getBucketNumber(), 10),
+                ImmutableList.of(getInternalNode("1"), getInternalNode("2")),
+                IntStream.range(0, 10).mapToObj(TestPartitionHandle::new).collect(toImmutableList()),
+                OptionalInt.of(1));
+    }
+
+    private static InternalNode getInternalNode(String id)
+    {
+        return new InternalNode(id, URI.create(id), new NodeVersion("test"), false);
+    }
+
+    private static class TestSplit
+            implements ConnectorSplit
+    {
+        private final int bucketNumber;
+
+        private TestSplit(int bucketNumber)
+        {
+            this.bucketNumber = bucketNumber;
+        }
+
+        public int getBucketNumber()
+        {
+            return bucketNumber;
+        }
+
+        @Override
+        public boolean isRemotelyAccessible()
+        {
+            return false;
+        }
+
+        @Override
+        public List<HostAddress> getAddresses()
+        {
+            return ImmutableList.of();
+        }
+
+        @Override
+        public Object getInfo()
+        {
+            return this;
+        }
+    }
+
+    private static class TestPartitionHandle
+            extends ConnectorPartitionHandle
+    {
+        private final int bucket;
+
+        public TestPartitionHandle(int bucket)
+        {
+            this.bucket = bucket;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof TestPartitionHandle)) {
+                return false;
+            }
+            TestPartitionHandle that = (TestPartitionHandle) o;
+            return bucket == that.bucket;
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(bucket);
+        }
+    }
+
+    public class TestSourceScheduler
```

Comment:
Do we usually say `TestSourceScheduler` or `TestingSourceScheduler` ?  cc @nezihyigitbasi

Reply:
I think it should be `TestingFoo`. Thanks for pointing this out.